### ISSUE TITLE
Multi-Device support requirements for semaphores

### ISSUE THREAD

**Issue by nikhiljnv** [KhronosGroup/OpenCL-Docs#973]
Came up as part of discussion on #944 and #957 
Tracking issue to discuss need and coverage for multi-device support for semaphores.

**Comment by nikhiljnv** [comment]
Some high-level questions -

- State of semaphore across devices
- Default device support for multi-device context in the absence of specific device list
- Ability to pass subset of devices at the time of semaphore creation (#957)

**Comment by bashbaug** [comment]
Discussed a few interesting scenarios in the October 3rd memory subgroup:

1. With a multi-device context, would we want to signal a semaphore from a command-queue for one device and wait on the semaphore from a command-queue on a different device (both devices in the same context)?
2. Would any implementations generate more efficient code if they knew that a semaphore created against a multi-device context would only be used by one of the devices, say by signaling the semaphore from one command-queue and waiting on the semaphore from a different command-queue on the same device?

**Comment by bcalidas** [comment]
> Discussed a few interesting scenarios in the October 3rd memory subgroup:
> 
> 1. With a multi-device context, would we want to signal a semaphore from a command-queue for one device and wait on the semaphore from a command-queue on a different device (both devices in the same context)?

 Yes we would. 

> 2. Would any implementations generate more efficient code if they knew that a semaphore created against a multi-device context would only be used by one of the devices, say by signaling the semaphore from one command-queue and waiting on the semaphore from a different command-queue on the same device?

>> That's possible

As it stands it appears that if an implementation supports semaphores then it must support semaphores on all the devices and it should be possible to create a semaphore that works on any combination of devices.

To avoid this implication, we could restrict semaphores to a single device only and force export + import in order to have signaling and waiting across queues that target different devices in a context. This would make both the API and implementations more robust.

**Comment by bcalidas** [comment]
After further internal discussion, we feel that restricting semaphores to a single device brings clarity to the API and to the usage of semaphores. It also makes the implementation much easier.
This would require an update to the base semaphore spec and CTS.

**Comment by bcalidas** [comment]
Following up on the discussion from 10 Oct 2023. There was a general feeling that restricting semaphores to a single device was the right approach. This means that the base semaphore spec will need to be updated since a device handle list specifying exactly one device must always be passed in as one of the properties in clCreateSemaphoreWithPropertiesKHR.

With the current specs, applications wishing to use semaphores across 2 devices in the same context, must use explicit export and import of semaphore handles to signal and wait.  

Going forward, we could add a command along the lines of clLinkSemaphores (sem_a, sem_b ) to make signaling and waiting across 2 different devices in the same context easier for an application to implement. It's possible that at the moment, we don't have any implementation that needs this functionality. If this is the case, clLinkSemaphores() could be defined in a follow-up extension when such functionality can be verified.

**Comment by nikhiljnv** [comment]
As discussed on Memory subgroup call on October 17, 2023, we restrict the semaphore usage to single device and defer multi-device support to a future extension.

**Comment by nikhiljnv** [comment]
Issues to resolve -
1. Should CL_SEMAPHORE_DEVICE_HANDLE_LIST_KHR be required for single device context?
2. What should be the behavior in case CL_SEMAPHORE_DEVICE_HANDLE_LIST_KHR is not specified? (For single device context vs multi-device context)

**Comment by nikhiljnv** [comment]
Just capturing thoughts based on the discussion during OpenCL Memory subgroup call on October 17, 2023.
Possible alternatives discussed
1. Specifying CL_SEMAPHORE_DEVICE_HANDLE_LIST_KHR  can be optional for single device context as there is only one device anyways. However, CL_SEMAPHORE_DEVICE_HANDLE_LIST_KHR  must be specified to point to a single device that semaphore will be usable on in case of multi-device context and not specifying this should be an error. This allows existing code to continue to work in case of single device context without any changes, but may start failing for multidevice context.
2. Specifying CL_SEMAPHORE_DEVICE_HANDLE_LIST_KHR   should always be mandatory irrespective of number of devices in the context. Would break existing code in case of single device context and will require changes for it to work, but gives more deterministic behavior going forward.

A couple of related questions that came up during the discussion -
1. How does any of the decision above affect us in future when we may have layered extension that supports multi-device semaphores?
2. How does this affect other similar scenarios (e.g. device USM) and if we need to be consistent?

**Comment by nikhiljnv** [comment]
Just capturing summary of discussion from memory subgroup call on November 7, 2023.
1. Single device context, no device handle list should work without any error.
2. Single device context with a device handle list that specifies the device that is in the context should work without any error.
3. Multi device context without device handle list should throw an error. 
4. Multi device context with device handle list pointing to just one device should still work. 
5. Single/Multi device context with device handle list pointing to more than one devices should throw an error.
6. By default semaphore object would belong to all devices in the context unless a subset of devices is specified by device handle list. cl_khr_semaphore only support semaphore on a single device.

Need to discuss behavior in case of subdevices that are created from the same parent device. In general, may need more discussion the behavior of certain APIs in case of subdevices (Tracked on separate issue).

**Comment by bcalidas** [comment]
We pushed a PR  that updates the spec language for one semaphore per device.
https://github.com/KhronosGroup/OpenCL-Docs/pull/996/
